clippy_utils/src/ty.rs

```diff
@@ -1290,3 +1242,101 @@ pub fn normalize_with_regions<'tcx>(tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>
 pub fn is_manually_drop(ty: Ty<'_>) -> bool {
     ty.ty_adt_def().map_or(false, AdtDef::is_manually_drop)
 }
+
+#[derive(Clone, Copy)]
+pub enum Sizedness {
+    /// The type is uninhabited. (e.g. `!`)
+    Uninhabited,
+    /// The type is zero-sized.
+    Zero,
+    /// The type has some other size or an unknown size.
+    Other,
+}
+impl Sizedness {
+    pub fn is_zero(self) -> bool {
+        matches!(self, Self::Zero)
+    }
+
+    pub fn is_uninhabited(self) -> bool {
+        matches!(self, Self::Uninhabited)
+    }
+}
+
+/// Calculates the sizedness of a type.
+pub fn sizedness_of<'tcx>(tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>, ty: Ty<'tcx>) -> Sizedness {
+    fn for_list<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        tys: impl IntoIterator<Item = Ty<'tcx>>,
+    ) -> Sizedness {
+        let mut res = Sizedness::Zero;
+        for ty in tys {
+            match sizedness_of(tcx, param_env, ty) {
+                Sizedness::Uninhabited => return Sizedness::Uninhabited,
+                Sizedness::Other => res = Sizedness::Other,
+                Sizedness::Zero => {},
+            }
+        }
+        res
+    }
+
+    match *ty.kind() {
+        ty::FnDef(..) => Sizedness::Zero,
+        ty::Tuple(tys) => for_list(tcx, param_env, tys),
+        ty::Array(_, len) if len.eval_target_usize(tcx, param_env) == 0 => Sizedness::Zero,
```

Comment:
Won't this ICE if len is a generic parameter?
```rs
type M<const N: usize> = HashMap<(), [(); N]>;
```

Reply:
Fixed.